ttnn/cpp/ttnn/tensor/tensor_layout.cpp

```diff
@@ -0,0 +1,358 @@
+//    to pad between Channels and Batches with a row granularity.
+// Examples:
+// 1. Given LegacyPaddedShape (1, 1, 16, 20), alignment is (1, 1, 16, 20)
+//    means Shape (1, 1, 16, 16) which in theory takes physical space (1 * 1 * 16, 16)
+//    will take (1 * 1 * 16, 20)
+//
+// 2. Given LegacyPaddedShape (1, 1, 32, 32), alignment is (1, 1, 32, 32)
+//    Shape (1, 1, 16, 16)
+//      which in theory takes physical space (1 * 1 * 16, 16)
+//      will take (1 * 1 * 32, 32)
+//
+// 3. Given LegacyPaddedShape (2, 3, 32, 32), alignment is (2, 3*32, 32, 32).
+//    Shape (2, 2, 16, 16)
+//      which in theory takes physical space (2 * 2 * 16, 16)
+//      will take (2 * 3 * 32, 32)
+
+Alignment legacyPaddedShapeToAlignment(const ttnn::SimpleShape& legacyPaddedShape) {
+    std::vector<uint32_t> values;
+    const auto rank = legacyPaddedShape.rank();
+    values.resize(rank);
+
+    if(rank > 1) {
+        values[rank - 1] = legacyPaddedShape[rank - 1];
+    }
+    if(rank > 2) {
+        values[rank - 2] = legacyPaddedShape[rank - 2];
+    }
+    for (size_t i = rank - 3; i >= 0; i--) {
+        values[i] = legacyPaddedShape[i] * values[i + 1];
+    }
+
+    Alignment result(values);
+    return result;
+}
+}
+
+// Private constructor to create TensorLayout from LegacyPaddedShape
+TensorLayout::TensorLayout(DataType dataType, Layout layout, const MemoryConfig& memoryConfig, const ttnn::SimpleShape& legacyPaddedShape)
+    : TensorLayout(dataType, layout, memoryConfig, legacyPaddedShapeToAlignment(legacyPaddedShape)) {
+    mLegacyPaddedShape = legacyPaddedShape;
+}
+
+TensorLayout TensorLayout::fromLegacyPaddeShape(DataType dataType, Layout layout, const MemoryConfig& memoryConfig, const ttnn::SimpleShape& legacyPaddedShape) {
+    return TensorLayout(dataType, layout, memoryConfig, legacyPaddedShape);
+}
+
+void TensorLayout::initializeAlignment() {
+    if(mAlignment.size() != 0)
+        return;
+
+    switch(mLayout) {
+        case Layout::ROW_MAJOR: {
+            mAlignment = Alignment({sizeof(uint32_t) % element_size_bytes()});
+            break;
+        }
+        case Layout::TILE: {
+            mAlignment = Alignment({mTileSize.height(), mTileSize.width()});
+            break;
+        }
+        case Layout::INVALID:
+            TT_THROW("Invalid Layout");
+    }
+}
+
+void TensorLayout::validateCustomAlignment() const
+{
+    if(mAlignment.size() == 0)
+        return;
+
+    switch(mLayout) {
+        case Layout::ROW_MAJOR: {
+            if(mAlignment.size() > 0) {
+                auto widthAlignment = mAlignment[-1];
+                auto element_size = element_size_bytes();
+                auto page_alignment = sizeof(uint32_t) % element_size;
+                TT_FATAL(widthAlignment % page_alignment == 0,
+                "Wrong custom Tensor Layout alignment {}. For Row Major layout with element size {}bytes the innermost dimension must align to {}. This is because Buffer data is packes as uint32_t (4 bytes).",
```

Comment:
I think we should always perform this check. If `mAlignment.size() == 0` we should assume `widthAlignment` to be equal one